### ISSUE TITLE
Lab10 Review - Claudio Savelli (S317680)

### ISSUE THREAD

**Issue by ClaudioSavelli** [loregrc/Computational-intelligence#4]
Hello again! I saw your Lab10 and was very interested, so I decided to review your work! 

Starting with the writing and structure of the code, I found it excellent to divide each method into its own cell with a title, making easy to understand the purpose each function had within your workflow. The only thing I have to say about this is that it would probably have been preferable to write general functions that were valid whether the agent played first or second (or both), thus avoiding writing a lot of unnecessary code and making the work less verbose. This could easily be achieved by passing the two players to `intelligent_game` (rather than `intelligent_game_X` or `intelligent_game_O`). 

As for the experiments you carried out, I found them very interesting and really complete. I appreciated how you put the whole workflow you followed inside the README, pointing out its strengths and weaknesses, making the analysis of the work extremely more immediate and clear, so much so that it seemed as if I were working with you! 

Another plus point is that you not only limited yourself to optimising the agents against a random player but also trained them to play against optimal tic-tac-toe players, forcing a draw every time, even going beyond the laboratory objectives!

The only note I want to make about the experiments is the number of games played during the training phase. In fact, you could have considered, instead of increasing the number of training games from 100,000 to 500,000, which also means slowing down the training time, exploiting the nature of the tic-tac-toe table! In fact, for each position in the play, there can be up to 4 totally identical positions by simply rotating the board (unless the first move is 'X' in the middle). In this way you don't only decrease the number of games to be played by a factor of 3 but also ensures that you see more states in large numbers so that the value dictionary covers all possible positions more evenly.

In conclusion, I think the work is superb, perfectly covering not only the required objectives but also going beyond them, with a neat code accompanied by an excellent description. Great job!
